sdk/metric/sdk.go

```diff
@@ -115,16 +119,143 @@ type (
 		recorder export.Aggregator
 	}
 
+	observerResult struct {
+		observer *observer
+	}
+
+	int64ObserverResult struct {
+		result observerResult
+	}
+
+	float64ObserverResult struct {
+		result observerResult
+	}
+
+	observerCallback func(result observerResult)
+
+	observer struct {
+		// callback has to be aligned for 64-bit atomic operations.
+		callback unsafe.Pointer // *observerCallback
+
+		meter      *SDK
+		descriptor *export.Descriptor
+		// recorders maps encoded labelset to the pair of
+		// labelset and recorder
+		recorders map[string]labeledRecorder
+	}
+
+	labeledRecorder struct {
+		recorder      export.Aggregator
+		labels        *labels
+		modifiedEpoch int64
+	}
+
+	int64Observer struct {
+		observer *observer
+	}
+
+	float64Observer struct {
+		observer *observer
+	}
+
 	ErrorHandler func(error)
 )
 
 var (
-	_ api.Meter               = &SDK{}
-	_ api.LabelSet            = &labels{}
-	_ api.InstrumentImpl      = &instrument{}
-	_ api.BoundInstrumentImpl = &record{}
+	_ api.Meter                 = &SDK{}
+	_ api.LabelSet              = &labels{}
+	_ api.InstrumentImpl        = &instrument{}
+	_ api.BoundInstrumentImpl   = &record{}
+	_ api.Int64Observer         = int64Observer{}
+	_ api.Float64Observer       = float64Observer{}
+	_ api.Int64ObserverResult   = int64ObserverResult{}
+	_ api.Float64ObserverResult = float64ObserverResult{}
 )
 
+func (r observerResult) observe(number core.Number, ls api.LabelSet) {
+	r.observer.recordOne(number, ls)
+}
+
+func (o *observer) recordOne(number core.Number, ls api.LabelSet) {
+	if err := aggregator.RangeTest(number, o.descriptor); err != nil {
+		o.meter.errorHandler(err)
+		return
+	}
+	recorder := o.getRecorder(ls)
+	if recorder == nil {
+		// The instrument is disabled according to the
+		// AggregationSelector.
+		return
+	}
+	if err := recorder.Update(context.Background(), number, o.descriptor); err != nil {
+		o.meter.errorHandler(err)
+		return
+	}
+}
+
+func (o *observer) getRecorder(ls api.LabelSet) export.Aggregator {
+	labels := o.meter.labsFor(ls)
+	lrec, ok := o.recorders[labels.encoded]
+	if ok {
+		lrec.modifiedEpoch = o.meter.currentEpoch
+		o.recorders[labels.encoded] = lrec
+		return lrec.recorder
+	}
+	rec := o.meter.batcher.AggregatorFor(o.descriptor)
+	if o.recorders == nil {
+		o.recorders = make(map[string]labeledRecorder)
+	}
+	// TODO: This may store nil recorder in the map, thus
```

Comment:
This is intentionally done for the other instruments. We want a way to disable these, but you could be right about this: re-querying the selector API allows us to dynamically reconfigure it later. We're so behind on configuration work I'd be happy to enter a comment here saying this is intentional and revisit it later.

Reply:
I updated the comment then.